Add support for monero generate_key_derivation and subaddress/output …

vtnerd opened the pull request:
Reviewers, please look at updated README.md first which explains the goals of this PR in particular.

-------------------

On ryzen 3900x the speedup for `crypto::generate_key_derivation` is ~153% and the speedup for `crypto::derive_subaddress_public_key` is ~141%. When looking at the total crypto operations for a standard 2-output transaction, this results in a ~147% speedup on this same CPU. On a recentish Intel laptop the speedup for 2-output transactions is ~140%.

I have yet to do a speed comparison in wallet2 directly. I would not expect an exact ~140%-192% speedup because there are other factors limiting the scanning. The crypto operations are fairly dominant in the profile graphs though, so there will be decent gains. I will publish numbers at some point, unless it appears fairly quickly that this PR is going to be rejected.

----------------

@kenshi84 @hyc @stoffu @xiphon @moneromooo-monero there is somewhat advanced ASM in this PR. The ASM included here is _adapted_ from the existing ASM in `crypto_sign/ed25519/<variant>/choose_t.s`. That assembly was written where `z == 1` since the scalarmult tables could be pre-generated. The ASM here took the existing ASM and added support for the `z` value in the group element with an unused x86-64 register. Diffing against the originals will show this.

The other options: (1) drop constant-time nature possibly leaking viewkey in some contexts or (2) 7 additional field inversions when generating the table for scalarmult. This table is needed when doing the ECDH for every tx (ed25519 typically does not have ECDH against arbitrary points, only `G`). Pinging @SarangNoether @surae-noether too.

---------------

@moneromooo-monero @moneroexamples This was specifically designed to be usable in outside projects. I _believe_ openmonero and monero-light-wallet-server can drop the dependency on internal `monero-project/monero` libraries if they use ZMQ-JSON-RPC and add: (1) keccak, the (2) custom varint serialization, and (3) support for parsing `tx_extra`. The first two are required for `crypto::derivation_to_scalar` which is not provided by these sets of changes. The **other** big blocker for those projects is portability - @moneroexamples please let me know if you are interested and we can gauge support for adding `ref10` to this repo. This will not be used by the default monero-wallet, but would provide portability for the wallet servers.

The other advantage to outside projects is all of this code should be easier to audit than the internal monero crypto library - the ref10 library should require zero changes and these amd64 variants only had position-independent changes to the ASM.

vtnerd committed:
Add support for monero generate_key_derivation and subaddress/output step

SarangNoether commented:
^ Surae Noether is @b-g-goodell here.

vtnerd commented:
I mistakenly ran the benchmarks with randomx mining on my 3900x whoops. The numbers are ~141%, ~153% and ~147% above which is closer to the laptop in percentage increase (overall the timing is much faster). Updated the comment above.

moneroexamples commented:
@vtnerd 

I don't have anything against this. But I don't know when I would be able to adapt openmonero to use it. 

p.s.
Awesome job, btw.

vtnerd commented:
@moneroexamples I'm glad your initial reaction was positive. I primarily wanted to make you aware of this effort because there are significant performance advantages and ease-of-use if `ref10` is added. Let me know if there is something that can accommodate your usage.

Another thing I didn't mention - I can add support for `amd64-64-24k`, `amd64-51-30k`, and `ref10` ECDH "precomp" (if reviewers were willing). There isn't any benefit to the default monero wallet, but these servers can amortize the cost of the first ECDH step by pre-computing the table once for N viewkeys. The time savings is x% **per-viewkey - 1** (I forget the exact amount at the moment).

vtnerd commented:
I was able to implement the two functions described in the `README.md`, using tweetnacl as the base in 301 lines of C code.

@b-g-goodell @SarangNoether @moneromooo-monero would be it easier to drop the amd64 stuff in this PR, and provide just the CMake + tweetnacl version for now? It would limit the review to just the library concept with a minimal C base implementation that can be reviewed by the MRL.

vtnerd commented:
Pinging @knaccc , who may also be interested ...

knaccc commented:
@vtnerd Thanks, this sounds awesome. 

> (1) drop constant-time nature possibly leaking viewkey in some contexts

I'm trying to think of ways in which this leak could happen, it seems extremely low risk to me to drop constant-time ECDH blockchain scanning if it significantly improves performance

>  or (2) 7 additional field inversions when generating the table for scalarmult.

Please could you elaborate on what you mean here? What is it that is creating 7 additional inversions?

> I can add support for amd64-64-24k, amd64-51-30k, and ref10 ECDH "precomp" (if reviewers were willing)

I'd assume that being able to do precomp is very useful for doing a scalarmult with e.g. H, right?

vtnerd commented:
> I'm trying to think of ways in which this leak could happen, it seems extremely low risk to me to drop constant-time ECDH blockchain scanning if it significantly improves performance

The monero daemon and viewkey are not necessarily on the same machine or owner. Monero now provides a mechanism for finding publicly available RPC ports for instance. The usage is clearly less common, but I think its preferable to "do the same thing" here and keep the behavior consistent with the existing Monero crypto code. This does have to be weighed against the complexity of having to review/maintain ASM code that is close but slightly different to the upstream project.

> Please could you elaborate on what you mean here? What is it that is creating 7 additional inversions?

The arbitrary point scalarmult code (for use by the public-key in a transaction) generates a 3-bit "table" for performance purposes - it reduces the total point doublings needed. The size was chosen because (1) existing monero ref10 code creates a table of this size and (2) `sc25519_window4` provided by the amd64 implementations generates a "bit-array" for this size table.

The problem is the existing constant-time table selection ASM does not use projection (`z`) coordinates because only pre-generated tables for `G` are used by the ASM. The existing ASM can be used if each element in the runtime-generated table have the projection `z` coordinate normalized to `1`. This requires a field inversion and 3 field multiplications for 7 elements (the first element is always normalized to `z == 1`).

> I'd assume that being able to do precomp is very useful for doing a scalarmult with e.g. H, right?

Probably, but I haven't looked into that. We should be able to put a table in static-memory too, like with point-`G`.

It is also useful for doing scalarmult against a transaction public key when 2+ viewkeys have to be scanned against it. This occurs primarily in "light wallet" servers or any wallet tracking multiple primary accounts.

moneromooo-monero commented on src/amd64/amd64.c.inc:
```diff
@@ -0,0 +1,201 @@
+}
+
+// similar to loops in existing implementation, but uses dynamic table instead of fixed `G`.
+static void monero_scalarmult(ge25519_p3* r, ge25519_pniels const* base, char const* sec)
```
Maybe an unfortunate name since it does a mul8 at the end. static so does not matter that much I guess.

vtnerd replied:
I named it `monero_scalarmult` because its not-standard. But you are correct, this clashes with the naming scheme elsewhere. `scalarmult_cryptonote` ?

moneromooo-monero replied:
I was kinda thinking just add "8" or something along those lines.

luigi1111 replied:
Seems most similar to https://github.com/monero-project/monero/blob/master/src/crypto/crypto.cpp#L184

vtnerd replied:
Renamed, see updated version being pushed shortly.

moneromooo-monero reviewed (commented):
Can you please link to the actual source you imported for those two asm libs ?

vtnerd commented:
The "base"/"template" ASM for amd64-64-30k is [here](https://github.com/monero-project/supercop/blob/master/crypto_sign/ed25519/amd64-64-24k/choose_t.s) and for amd64-51-30k is [here](https://github.com/monero-project/supercop/blob/master/crypto_sign/ed25519/amd64-51-30k/choose_t.s).

moneromooo-monero commented:
Do you have a source that's from the actual author, not from the monero repo ? I wanted to compare the sources, and comparing with something not from the original defeats the point.

moneromooo-monero commented:
Hmm, that source is not what I thought it was. I mean this pulls the actual asm implementation of some ops, right ?

iamamyth commented on src/amd64/amd64.c.inc:
```diff
@@ -0,0 +1,201 @@
+    ge25519_pnielsadd_p1p1(&tp1p1, r, &t);
+    ge25519_p1p1_to_p2((ge25519_p2*)r, &tp1p1);
+
+    if (i == 0) break;
```
I would expect the implementation to be faster if the loop were unrolled to eliminate the `break` at 0 (i.e. decrease the iteration count by one and execute the three needed statements after the loop with a fixed index). Compilers usually don't figure out this variant of unrolling for you, though it's easy to check the generated assembly. I also wonder whether an explicit alignment for `b` would help; it should fit in a 64-byte cache line on Intel and AMD processors, but only if the compiler picks a suitable alignment.

vtnerd replied:
Gcc 9.2 with `-ofast` didn't unroll the loop either way. This isn't exactly a "hot loop" since many expensive functions are called from within it.

vtnerd commented:
The original code is available on [Daniel Bernstein's website](https://bench.cr.yp.to/supercop.html). There is also a well-known [github mirror](https://github.com/floodyberry/supercop). The `master` branch in this repo should be unchanged from those versions. The current `monero` branch should only have changes for relative position code.

moneromooo-monero commented on src/amd64/amd64-51-30k-choose_tp.s:
```diff
@@ -0,0 +1,2189 @@
+# qhasm: t = *(uint64 *)(basep + 184 + pos)
+# asm 1: movq   184(<basep=int64#4,<pos=int64#1),>t=int64#15
+# asm 2: movq   184(<basep=%rcx,<pos=%rdi),>t=%rbp
+movq   224(%rcx,%rdi),%rbp
```
Comparing to the original, the changes up till now are just subtracting 40. On this one, the comments points to the "new" offset, though the code matches the expected progression. Why is that ? Simple bug in the original ? If so, bug in the comment, or the actual code ?

vtnerd replied:
The bug is almost certainly in the comment. The wallet PR has a benchmark utility which first runs a diagnostic check ensuring that the existing output from the `crypto::...` functions is identical to these accelerated versions. An offset bug here would be multiplying the secret key bits incorrectly, and the probability of it matching repeatedly are pretty small.

I will double check to see what happened, most likely I just forgot to update the comment. Initially I wasn't updating the comments at all, but part way through I started updating them. Must've missed some on the re-pass back.

moneromooo-monero commented on src/amd64/amd64-51-30k-choose_tp.s:
```diff
@@ -0,0 +1,2189 @@
+# qhasm: t = *(uint64 *)(basep + 320 + pos)
+# asm 1: movq   320(<basep=int64#4,<pos=int64#1),>t=int64#15
+# asm 2: movq   320(<basep=%rcx,<pos=%rdi),>t=%rbp
+movq   320(%rcx,%rdi),%rbp
```
From this point, the offets change from -40 to -80. Why ?

vtnerd replied:
The code is touching every entry in a table (constant time), and conditionally moving the entry related to the bits set in the secret key. Each entry in the table now has an additional `z` field value, along with the 3 original field values, and each field value is 40 bytes (64bits*5limbs / 8bitsbyte). So after every entry in the table, the offset needed for the first original field value increases by 40. 

The [field element struct](https://github.com/monero-project/supercop/blob/0d6dff95350a20f6fe2bafe1ff6730458dd5ddef/crypto_sign/ed25519/amd64-51-30k/fe25519.h#L23), the [table entry struct for the original ASM](https://github.com/monero-project/supercop/blob/0d6dff95350a20f6fe2bafe1ff6730458dd5ddef/crypto_sign/ed25519/amd64-51-30k/ge25519.h#L61), and the [new table entry struct for this ASM](https://github.com/monero-project/supercop/blob/0d6dff95350a20f6fe2bafe1ff6730458dd5ddef/crypto_sign/ed25519/amd64-51-30k/ge25519.h#L68) just below it. The last `struct` has a `z` parameter which allows field inversions to be "batched" once, after the main processing loop. The original never had this because ed25519 signing uses a fixed basepoint (`G`), so the tables were pre-computed with every `z` value normalized to `1` for time/space savings.

EDIT: Clarification, the table was basically `ge25519_niels foo[8]` and now its `ge25519_pniels foo[8]`, so the offset from the original gets bumped in the exact spot where the `z` value was added to the struct.

moneromooo-monero commented on src/amd64/amd64-51-30k-choose_tp.s:
```diff
@@ -0,0 +1,2189 @@
+# qhasm: t = *(uint64 *)(basep + 272 + pos)
+# asm 1: movq   272(<basep=int64#4,<pos=int64#1),>t=int64#15
+# asm 2: movq   272(<basep=%rcx,<pos=%rdi),>t=%rbp
+movq   352(%rcx,%rdi),%rbp
```
This one also has the "new" offset in comments. Given I have no idea what data this is trying to copy, I can't guess at what the bug is.

moneromooo-monero replied:
And the same oddities continue later, I'm not pointing them out.

moneromooo-monero commented on src/amd64/amd64-64-24k.c:
```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2017, The Monero Project
```
I'm guessing the monero specific code is new ?

vtnerd replied:
Changed the copyrights to 2020 in both of these files.

moneromooo-monero commented on src/amd64/amd64-64-24k.c:
```diff
@@ -0,0 +1,73 @@
+/* taken from loop in ed25519/amd64-64-24k/ge25519_double_scalarmult.c */
+static void ge25519_p1p1_to_pniels(ge25519_pniels* out, ge25519_p1p1 const* in)
+{
+  assert(out && in);
```
For code called by wallets, I've tried to replace assert with _exit when NDEBUG, to avoid cores being dumped via SIGABRT (thereby dumping secrets to disk). git grep local_abort.

vtnerd replied:
I'm not sure why I had an `assert` in just this one location. Seems like either more functions should have an assert or none. I simply dropped this `assert` for now.

moneromooo-monero commented on src/amd64/amd64.c.inc:
```diff
@@ -0,0 +1,201 @@
+  /* 5. Choose the desired square root according to parity:
+  */
+  if(fe25519_getparity(&r->x) == (1-par)) // only change from original function was `!=` -> `==`
+    fe25519_neg(&r->x, &r->x);
```
The difference from the original here is:

```
   /* 5. Choose the desired square root according to parity:
   */
-  if(fe25519_getparity(&r->x) != (1-par))
+  if(fe25519_getparity(&r->x) == (1-par)) // only change from original function was `!=` -> `==`
     fe25519_neg(&r->x, &r->x);
 
   fe25519_mul(&r->t, &r->x, &r->y);
```

No clue why, so I'm putting this here and will ask luigi.

vtnerd replied:
@luigi1111 

The Ed25519 implementations unpack and negate the point at the same time. This is because public-keys are only used to [verify a signature](https://github.com/monero-project/supercop/blob/0d6dff95350a20f6fe2bafe1ff6730458dd5ddef/crypto_sign/ed25519/amd64-64-24k/open.c), which is an implementation of the [EdDSA signature equation](https://en.wikipedia.org/wiki/EdDSA). The function is subtracting the value on the far right from the left side of the equation to compare equality to the point `R`.

This function isn't strictly needed, but the x-coordinate would need to be immediately negated before calculating the ECDH for the transaction, etc. So this is for optimization purposes.

luigi1111 replied:
@vtnerd ah, thanks.

moneromooo-monero commented:
I'm happy with the PR, though I don't have the crypto knowledge to review amd64.c.inc. Hopefully luigi does that.

vtnerd commented:
This is annoying. [Github's compare function seems to be broken](https://github.com/monero-project/supercop/compare/69e6655...07beaba) - that's definitely the before-and-after commit hashes.

Changes:
- Renamed `monero_scalarmult` to `scalarmult_p1p1`. This variant does not do the mul8 at end.
- Added internal private function `scalarmult` which does standard ECDH on ed25519.
- Added public function(s) `monero_crypto_ge25519_scalarmult` which uses new private function.
- Fixed comments mentioned incorrect offsets.
- Fixed copyrights on two files.
- Removed `assert` in one location.

moneromooo-monero reviewed (approved):
Looks like just changes to a documentation, I suspect that's what vtnerd refers to. Github might have lost a commit in there.
